### PR TITLE
NFS client

### PR DESCRIPTION
**Pull request by mvollmer**
- [x] Also show mounts not in fstab
- [x] Create and remove mount point directories
- [x] Avoid flicker of arming machinery when list is rendered
- [x] Use showmount to help with/validate remote bits
- [x] Avoid quadratic complexity when preparing mount report
- [x] Show nice text when there are no nfs mounts
- [x] Tests
- [x] Use "mount -o remount" if only options change
- [x] Layout when armed
- [x] Don't let the user change server/remote/dir when mount is busy.
- [x] Refuse to remove busy mounts.
- [x] Reimplement lsof/fuser withing nfs-mounts.py
- [x] Let user kill all systemd units that keep a mount busy
- [x] Use python3 in preference to python2
- [x] Add nfs-server and python3-dbus to Debian/Ubuntu images
- [x] Make video for release notes
- [x] #7852 
- [x] #7812 
- [x] #7827 

Video: https://www.youtube.com/watch?v=PGCBda3Le9Y

### CODE REVIEW

**Comment by mvollmer** [comment]
Screencast of this: https://www.youtube.com/watch?v=ov5kVCbqqw8&feature=youtu.be

**Comment by mvollmer** [comment]
> The Mount Options UI will be very error prone (I don't even know what input it wants), so that will need some work, as outlined in your checklist.

Yeah.  We should also be looking at normal filesystem mount options.  THese are just a string right now as well.  GNOME Disks has an elaborate dialog, which we can take as inspiration.

**Comment by andreasn** [comment]
Let's make the title of the unmount error dialog "Could not unmount the filesystem" instead of the generic "Error". Discussed briefly with Marius that it would be good to also list the process/processes that currently uses the mount and some way of killing those from the dialog itself.

**Comment by mvollmer** [comment]
> Discussed briefly with Marius that it would be good to also list the process/processes that currently uses the mount and some way of killing those from the dialog itself.

Let's split that out: https://trello.com/c/oRZ6D7Wy/548-nfs-client-kill-for-unmount

**Comment by mvollmer** [comment]
> Let's split that out: https://trello.com/c/oRZ6D7Wy/548-nfs-client-kill-for-unmount

Let's not. :-)

**Comment by cockpituous** [comment]
image-refresh in progress on cockpit-tests-p8kdw.
Log: http://fedorapeople.org/groups/cockpit/logs/image-refresh-7580-20170920-072727/

**Comment by cockpituous** [comment]
Task aborted.

**Comment by cockpituous** [comment]
image-refresh in progress on cockpit-tests-h2qdv.
Log: http://fedorapeople.org/groups/cockpit/logs/image-refresh-7580-20170920-075454/

**Comment by cockpituous** [comment]
image-refresh in progress on cockpit-tests-569rm.
Log: http://fedorapeople.org/groups/cockpit/logs/image-refresh-7580-20170920-075454/

**Comment by cockpituous** [comment]
image-refresh in progress on cockpit-tests-4z182.
Log: http://fedorapeople.org/groups/cockpit/logs/image-refresh-7580-20170920-075353/

**Comment by cockpituous** [comment]
image-refresh in progress on cockpit-tests-51j91.
Log: http://fedorapeople.org/groups/cockpit/logs/image-refresh-7580-20170920-075454/

**Comment by cockpituous** [comment]
image-refresh debian-stable done: https://github.com/cockpituous/cockpit/commits/image-refresh-debian-stable-20170920-082929

**Comment by cockpituous** [comment]
image-refresh in progress on cockpit-tests-vlgmj.
Log: http://fedorapeople.org/groups/cockpit/logs/image-refresh-7580-20170920-102121/

**Comment by cockpituous** [comment]
image-refresh in progress on cockpit-tests-qms4t.
Log: http://fedorapeople.org/groups/cockpit/logs/image-refresh-7580-20170920-102121/

**Comment by cockpituous** [comment]
image-refresh ubuntu-stable done: https://github.com/cockpituous/cockpit/commits/image-refresh-ubuntu-stable-20170920-105454

**Comment by cockpituous** [comment]
image-refresh in progress on cockpit-tests-qms4t.
Log: http://fedorapeople.org/groups/cockpit/logs/image-refresh-7580-20170920-105555/

**Comment by cockpituous** [comment]
image-refresh ubuntu-1604 done: https://github.com/cockpituous/cockpit/commits/image-refresh-ubuntu-1604-20170920-112323

**Comment by cockpituous** [comment]
image-refresh in progress on cockpit-tests-8b3kn.
Log: http://fedorapeople.org/groups/cockpit/logs/image-refresh-7580-20170920-112525/

**Comment by mvollmer** [comment]
New video: https://www.youtube.com/watch?v=jYRxAtq8wcA

**Comment by cockpituous** [comment]
Task failed: http://fedorapeople.org/groups/cockpit/logs/image-refresh-7580-20170920-112525/

**Comment by martinpitt** [comment]
@mvo: the debian-testing failure is because of a version mismatch between the [linux-latest](https://tracker.debian.org/pkg/linux-latest) metapackage and [linux](https://tracker.debian.org/pkg/linux) in debian testing, so `linux-headers-amd64` is currently uninstallable. A quick hack would be to avoid the metapackage and install the latest kernel and corresponding headers by hardcoded version:

```diff
diff --git a/bots/images/scripts/debian-testing.bootstrap b/bots/images/scripts/debian-testing.bootstrap
index be2b9aba..b05796ce 100755
--- a/bots/images/scripts/debian-testing.bootstrap
+++ b/bots/images/scripts/debian-testing.bootstrap
@@ -1,4 +1,4 @@
 #! /bin/sh -ex
 ARCH=x86_64
 DEBIAN_LATEST=$(virt-builder -l  | grep "$ARCH" | sort -r | grep -m1 '^debian-' | cut -d' ' -f1)
-exec $(dirname $0)/lib/debian.bootstrap "$1" "$2" "$DEBIAN_LATEST" "deb http://deb.debian.org/debian testing main"
+exec $(dirname $0)/lib/debian.bootstrap "$1" "$2" "$DEBIAN_LATEST" "deb http://deb.debian.org/debian testing main" "eatmydata apt install -y linux-image-4.12.0-2-amd64 linux-headers-4.12.0-2-amd64"
diff --git a/bots/images/scripts/lib/debian.bootstrap b/bots/images/scripts/lib/debian.bootstrap
index 17a44273..1adc5bb7 100755
--- a/bots/images/scripts/lib/debian.bootstrap
+++ b/bots/images/scripts/lib/debian.bootstrap
@@ -31,4 +31,5 @@ virt-builder $virt_builder_image \
              --run-command "sed -i 's/GRUB_TIMEOUT.*/GRUB_TIMEOUT=0/; /GRUB_CMDLINE_LINUX=/ s/"'"'"$/ console=ttyS0,115200 net.ifnames=0 biosdevname=0"'"'"/' /etc/default/grub" \
              --run-command "update-grub" \
              --run-command "sed -i 's/ens[^[:space:]]*/eth0/' /etc/network/interfaces" \
-             --run-command "export DEBIAN_FRONTEND=noninteractive; apt-get -y update; apt-get -y install eatmydata; eatmydata apt-get -y dist-upgrade"
+             --run-command "export DEBIAN_FRONTEND=noninteractive; apt-get -y update; apt-get -y install eatmydata; eatmydata apt-get -y dist-upgrade" \
+             --run-command "${5:-true}"

```
But that will only be valid for a few days, then we have to drop it again (which is fine, the next debian-testing image rebuild will tell us, and it's easy enough to remove).

**Comment by mvollmer** [comment]
> so linux-headers-amd64 is currently uninstallable.

I think we can wait a few days and see whether this clears up.

**Comment by andreasn** [comment]
As mentioned on IRC, the Unmount button is very long, so I suggest moving some of that information into the dialog text itself.
https://raw.githubusercontent.com/cockpit-project/cockpit-design/d389723574b294452b8ec193c63bceea3eeff13d/storage/nfs-unmount.png

**Comment by mvollmer** [comment]
> the Unmount button is very long, so I suggest moving some of that information into the dialog text itself.

Done.  I have also used half of the to existing 'teardown' mechanism for this.  As a later step, we can complete this and get the same kind of warning/behavior in other places where it applies.

**Comment by andreasn** [comment]
> As a later step, we can complete this and get the same kind of warning/behavior in other places where it applies.

Sounds good!

**Comment by mvollmer** [comment]
Ubuntu images need to be made again.

**Comment by cockpituous** [comment]
image-refresh in progress on cockpit-tests-jttb7.
Log: http://fedorapeople.org/groups/cockpit/logs/image-refresh-7580-20170922-092828/

**Comment by cockpituous** [comment]
image-refresh ubuntu-stable done: https://github.com/cockpituous/cockpit/commits/image-refresh-ubuntu-stable-20170922-101111

**Comment by cockpituous** [comment]
image-refresh in progress on cockpit-tasks-8wmgk.
Log: http://fedorapeople.org/groups/cockpit/logs/image-refresh-7580-20170922-101515/

**Comment by cockpituous** [comment]
image-refresh ubuntu-1604 done: https://github.com/cockpituous/cockpit/commits/image-refresh-ubuntu-1604-20170922-105959

**Comment by cockpituous** [comment]
image-refresh in progress on cockpit-tasks-kr33g.
Log: http://fedorapeople.org/groups/cockpit/logs/image-refresh-7580-20170925-114141/

**Comment by cockpituous** [comment]
image-refresh in progress on cockpit-tasks-pnww5.
Log: http://fedorapeople.org/groups/cockpit/logs/image-refresh-7580-20170925-114141/

**Comment by cockpituous** [comment]
image-refresh in progress on cockpit-tests-vgsh5.
Log: http://fedorapeople.org/groups/cockpit/logs/image-refresh-7580-20170925-114242/

**Comment by cockpituous** [comment]
image-refresh debian-testing done: https://github.com/cockpituous/cockpit/commits/image-refresh-debian-testing-20170925-122222

**Comment by cockpituous** [comment]
image-refresh in progress on cockpit-tasks-7ndtn.
Log: http://fedorapeople.org/groups/cockpit/logs/image-refresh-7580-20171004-073939/

**Comment by cockpituous** [comment]
image-refresh ubuntu-stable done: https://github.com/cockpituous/cockpit/commits/image-refresh-ubuntu-stable-20171004-081313

**Comment by cockpituous** [comment]
image-refresh in progress on cockpit-tasks-7ndtn.
Log: http://fedorapeople.org/groups/cockpit/logs/image-refresh-7580-20171004-081313/

**Comment by cockpituous** [comment]
image-refresh ubuntu-1604 done: https://github.com/cockpituous/cockpit/commits/image-refresh-ubuntu-1604-20171004-085656

**Comment by cockpituous** [comment]
image-refresh in progress on cockpit-tasks-zffgk.
Log: http://fedorapeople.org/groups/cockpit/logs/image-refresh-7580-20171004-090202/

**Comment by cockpituous** [comment]
image-refresh debian-stable done: https://github.com/cockpituous/cockpit/commits/image-refresh-debian-stable-20171004-093232

**Comment by cockpituous** [comment]
image-refresh in progress on cockpit-tasks-8czrx.
Log: http://fedorapeople.org/groups/cockpit/logs/image-refresh-7580-20171004-100202/

**Comment by cockpituous** [comment]
image-refresh ubuntu-stable done: https://github.com/cockpituous/cockpit/commits/image-refresh-ubuntu-stable-20171004-103838

**Comment by cockpituous** [comment]
image-refresh in progress on cockpit-tasks-2cqdg.
Log: http://fedorapeople.org/groups/cockpit/logs/image-refresh-7580-20171004-104343/

**Comment by cockpituous** [comment]
Task failed: http://fedorapeople.org/groups/cockpit/logs/image-refresh-7580-20171004-104343/

**Comment by cockpituous** [comment]
image-refresh in progress on cockpit-tasks-2xs49.
Log: http://fedorapeople.org/groups/cockpit/logs/image-refresh-7580-20171004-113030/

**Comment by cockpituous** [comment]
image-refresh debian-stable done: https://github.com/cockpituous/cockpit/commits/image-refresh-debian-stable-20171004-120202

**Comment by mvollmer** [comment]
Images are made in #7812

File: src/bridge/cockpitmountsamples.c [hunk 1]
```diff
@@ -57,9 +57,9 @@ cockpit_mount_samples (CockpitSamples *samples)
       if (strlen (line) == 0)
         continue;
 
-      /* Only look at real devices
+      /* Only look at real devices and remote mounts
        */
-      if (line[0] != '/')
+      if (line[0] != '/' && strchr(line, ':') == NULL)
```
Comment (martinpitt) [inline src/bridge/cockpitmountsamples.c]: I wonder if it is safe to stat remote file systems like this once or even twice (not sure about the frequency of our sampling) a second?

File: pkg/storaged/index.html [hunk 2]
```diff
@@ -844,7 +857,7 @@ <h4 class="modal-title">{{Title}}</h4>
             {{#Action}}
             {{#Danger}}<div class="modal-footer-danger">{{.}}</div>{{/Danger}}
             <div class="spinner spinner-sm pull-left" hidden></div>
-            <button class="btn btn-default" translatable="yes" data-dismiss="modal">Cancel</button>
+            <button class="btn btn-default" translatable="yes" data-action="cancel">Cancel</button>
```
Comment (martinpitt) [inline pkg/storaged/index.html]: This change is seemingly unrelated, and could at least do with an explanation in the commit log? Or even a separate commit? 5 lines down continues to use `data-dismiss`, and so does the error-popup.

File: pkg/storaged/client.js [hunk 3]
```diff
@@ -25,6 +25,9 @@
 
     var utils = require('./utils');
 
+    var nfs_mounts_py = require("raw!./nfs-mounts.py");
+    var pyinvoke = [ "sh", "-ec", "exec $(which python3 2>/dev/null || which python) $@", "--", "-" ];
```
Comment (martinpitt) [inline pkg/storaged/client.js]: One of these days this should really move to lib/, given that we copy this around more and more :-)

File: pkg/storaged/nfs-mounts.py [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+
+    def process(self, callback):
+        buf = os.read(self.fd, 4096)
+        pos = 0
+        while pos < len(buf):
+            (wd, mask, cookie, name_len) = struct.unpack('iIII', buf[pos:pos+16])
+            pos += 16
+            (name,) = struct.unpack('%ds' % name_len, buf[pos:pos + name_len])
+            pos += name_len
+            callback(wd, mask, name.decode().rstrip('\0'))
```
Comment (martinpitt) [inline pkg/storaged/nfs-mounts.py]: IMHO this code is too obscure and finnicky to carry it around multiple times in the source. Could the inotify bits move to pkg/lib/ and get included (statically during build) in both applications and storage? Of course a "proper" implementation in the bridge would be better in the long run (extending our `file.watch()` API), but we can't rely on that just yet on e. g. RHEL.

File: pkg/storaged/nfs-mounts.py [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+def parse_tab(name):
+    entries = [ ]
+    for line in open(name, "r").read().split("\n"):
+        sline = line.strip()
+        if sline == "" or sline[0] == "#":
+            continue
+        fields = list(map(field_unescape, re.split("[ \t]+", sline)))
+        if len(fields) > 0 and ":" in fields[0]:
+            entries.append(fields)
+    return entries
```
Comment (martinpitt) [inline pkg/storaged/nfs-mounts.py]: As you are already using ctypes, would it be possible/easier to use `{get,set,add}mntent()`? The `FILE*` pointers can just be passed around as opaque values, so they shouldn't hurt much, and `struct mntent` is just `char*` and `int` and thus should map to `ctypes` fairly well?

File: pkg/storaged/nfs-mounts.py [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+
+def process_fstab():
+    global fstab, fstab_by_remote
+    fstab = parse_tab("/etc/fstab")
+    fstab_by_remote = index_tab(fstab)
+
+def process_mtab():
+    global mtab, mtab_by_remote
+    mtab = parse_tab("/etc/mtab")
+    mtab_by_remote = index_tab(mtab)
```
Comment (martinpitt) [inline pkg/storaged/nfs-mounts.py]: `/etc/mtab` is really a broken-by-design concept and was abandoned years ago. Let's please not introduce new code for it. `/proc/self/mounts` should do fine.

File: pkg/storaged/nfs-mounts.py [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+            data.append({ "fstab": False, "fields": m, "mounted": True })
+    sys.stdout.write(json.dumps(data) + "\n")
+    sys.stdout.flush()
+
+def monitor():
+    process_mtab()
+    process_fstab()
+    report()
+    fstab_watcher = Watcher("/etc/fstab")
+    mtab_file = open("/proc/mounts", "r")
```
Comment (martinpitt) [inline pkg/storaged/nfs-mounts.py]: Let's be consistent and use `/proc/self/mounts` here, too. Like `/etc/mtab` itself, `/proc/mounts` is just a symlink and thus rather useless to monitor.

File: pkg/storaged/client.js [hunk 8]
```diff
@@ -404,6 +407,89 @@
...
+            return spawn_nfs_mounts([ "remove", JSON.stringify(entry) ]);
+        }
+
+        function mount_entry(entry) {
+            return spawn_nfs_mounts([ "mount", JSON.stringify(entry) ]);
+        }
+
+        function unmount_entry(entry) {
+            return spawn_nfs_mounts([ "unmount", JSON.stringify(entry) ]);
+        }
```
Comment (martinpitt) [inline pkg/storaged/client.js]: This could just call the `mount`/`umount` binaries directly, instead of going through the overhead of the Python execution? That also just calls these through subprocess, and not much else?

File: pkg/storaged/nfs-mounts.py [hunk 9]
```diff
@@ -0,0 +1,344 @@
...
+    modify_tab("/etc/fstab", lambda fields: new_fields if fields == old_fields else fields)
+    if entry["mounted"]:
+        if (new_fields[0] == old_fields[0]
+            and new_fields[1] == old_fields[1]
+            and new_fields[2] == old_fields[2]):
+            remount(new_fields)
+        else:
+            try:
+                unmount(entry)
+                rmdir_if_empty(old_fields[1])
```
Comment (martinpitt) [inline pkg/storaged/nfs-mounts.py]: Mount points don't have to be directories - you can also e. g. bind-mount a file onto another file. It is correct in this case to not remove the original file and for `rmdir_if_empty()` to fail, but that `errno.ENOTDIR` failure should be handled gracefully. (Not sure if something else here ensures that mount points are always directories).

File: pkg/storaged/nfs-mounts.py [hunk 10]
```diff
@@ -0,0 +1,344 @@
...
+        users(json.loads(argv[2]))
+
+try:
+    dispatch(sys.argv)
+except subprocess.CalledProcessError as e:
+    sys.exit(e.returncode)
+except:
+    sys.stderr.write("".join(traceback.format_exception_only(sys.exc_info()[0], sys.exc_info()[1])))
+    sys.stderr.flush()
+    sys.exit(1)
```
Comment (martinpitt) [inline pkg/storaged/nfs-mounts.py]: Is there much difference to the default exception handler? I'd rather prefer to not handle blanket unexpected exceptions, and  rather let things like Apport or ABRT kick in to catch and report these.

File: tools/cockpit.spec [hunk 11]
```diff
@@ -392,6 +392,8 @@ fi
 Summary: Cockpit user interface for storage, using Storaged
 Requires: %{name}-shell >= %{required_base}
 Requires: storaged >= 2.1.1
+Requires: python
+Requires: python-dbus
```
Comment (martinpitt) [inline tools/cockpit.spec]: Can we only do Python 2 for RHEL/CentOS, and use Python 3 on Fedora?

**Comment by mvollmer** [comment]
Depends on #7827

File: src/bridge/cockpitmountsamples.c (hunk 1 above)
Comment (mvollmer) [reply src/bridge/cockpitmountsamples.c]: We sample once a second.  What might go wrong if we sample too much?

File: pkg/storaged/client.js (hunk 3 above)
Comment (mvollmer) [reply pkg/storaged/client.js]: Yes.

In this case, the 'golden' solution would be to replace the whole Python script with new API in udisks2, I'd say.  I want to work on that as well.  We might decide that this PR here is only a proof-of-concept that we are not going to merge.  What do you think?

File: pkg/storaged/nfs-mounts.py (hunk 4 above)
Comment (mvollmer) [reply pkg/storaged/nfs-mounts.py]: I think we can require changes to the bridge for a new feature.  The shell and dashboard need to be compatible with all versions of bridges, but a page can always bring exactly the bridge it needs, no?

However, improving the bridge doesn't help a Python script.  We would have to reimplement it all in JavaScript and run it in the browser.  This is possible of course, but has no advantages except being able to use the bridge APIs.  (Writing synchronous code in Python is sooo much cleaner than writing asynch code in JavaScript, and we also avoid a ton of round trips.)

Which means, yes, I should move the inotify bits out into pkg/lib.

File: pkg/storaged/nfs-mounts.py (hunk 5 above)
Comment (mvollmer) [reply pkg/storaged/nfs-mounts.py]: Good point.

File: pkg/storaged/nfs-mounts.py (hunk 6 above)
Comment (mvollmer) [reply pkg/storaged/nfs-mounts.py]: Okay, let's use `/proc/self/mounts`.  The change notification depends on the file being a special file in `/proc`, so the old-school regular `/etc/mtab` file wouldn't work anyway.

(Out of interest, what is broken-by-design re /etc/mtab? Mount namespaces are kinda recent, no?)

File: pkg/storaged/nfs-mounts.py (hunk 7 above)
Comment (mvollmer) [reply pkg/storaged/nfs-mounts.py]: Yes.

File: pkg/storaged/client.js (hunk 8 above)
Comment (mvollmer) [reply pkg/storaged/client.js]: I think the overhead is small enough to be bearable and that we don't need to make this optimization.  This is not happening in some inner loop, after all. :-)

I prefer it this way for symmetry.  nfs-mounts.py exposes an API, and client.js talks to it.  nfs-mounts.py 'knows' how to call mount, and client.js doesn't need to.  In the future, I hope that udisks2 exposes the API.

File: pkg/storaged/nfs-mounts.py (hunk 9 above)
Comment (mvollmer) [reply pkg/storaged/nfs-mounts.py]: Good point, I didn't think of that.

File: pkg/storaged/client.js (hunk 8 above)
Comment (mvollmer) [reply pkg/storaged/client.js]: Also, there is not just "unmount", but also "stop-and-unmount".

File: pkg/storaged/client.js (hunk 3 above)
Comment (mvollmer) [reply pkg/storaged/client.js]: Done, #7852

File: src/bridge/cockpitmountsamples.c (hunk 1 above)
Comment (martinpitt) [reply src/bridge/cockpitmountsamples.c]: I'm not sure how much overhead that is. Imagine having a dozen NFS mounts which might not be super-fast, and sampling all of them once per second. That could create some network noise/latencies or even overlapping? E. g. nautilus/gvfs also doesn't continuously sample remote file systems, only on demand. So checking `df` on remote file systems once seems like an acceptable compromise?

Don't get me wrong, I don't veto this, but I wanted to point this out as a potential problem.

File: pkg/storaged/nfs-mounts.py (hunk 4 above)
Comment (martinpitt) [reply pkg/storaged/nfs-mounts.py]: Why does it need to happen in Java? The bridge already provides the inotify ↔ channel implementation, so that could be extended to be able to watch for directories too (what's done in Python). Then JS should only use the `file.watch()` API.

However:
> a page can always bring exactly the bridge it needs, no?

Not for RHEL 7.4 - there, the bridge is shipped in base and cannot (easily) change, but in extras we ship cockpit-{storaged,packagekit} which get rebased frequently. For that case we need to provide a compatibility shim if we want to support it there, at least until the rebase in 7.5. Or just don't support this feature until 7.5, which sounds perfectly acceptable, too.

File: pkg/storaged/nfs-mounts.py (hunk 6 above)
Comment (martinpitt) [reply pkg/storaged/nfs-mounts.py]: What's broken is that `/etc/mtab` contains information about all mounted file systems, but resides *on a mounted file system*. I. e. it had become stale in all sorts of situations when mounting stuff while the root fs was still read-only, doesn't work with permanent read-only root scenarios, and is just a layering violation by design.

File: pkg/storaged/nfs-mounts.py (hunk 10 above)
Comment (martinpitt) [reply pkg/storaged/nfs-mounts.py]: Oh, I was missing the `format_exception_only()`. Seems simpler to just use `except Exception as e:` and `sys.stderr.write(e)`?

File: src/bridge/cockpitmountsamples.c (hunk 1 above)
Comment (mvollmer) [reply src/bridge/cockpitmountsamples.c]: Yes, I think I am convinced.

File: pkg/storaged/nfs-mounts.py (hunk 4 above)
Comment (mvollmer) [reply pkg/storaged/nfs-mounts.py]: > Not for RHEL 7.4 - there, the bridge is shipped in base and cannot (easily) change, but in extras we ship cockpit-{storaged,packagekit} which get rebased frequently.

I see.  This is another argument against including remote mounts in our fsys sampling.

File: pkg/storaged/nfs-mounts.py (hunk 6 above)
Comment (mvollmer) [reply pkg/storaged/nfs-mounts.py]: Make sense, thanks!

File: pkg/storaged/nfs-mounts.py (hunk 10 above)
Comment (mvollmer) [reply pkg/storaged/nfs-mounts.py]: D'oh, I and was thinking "why is it so hard to make a shorter error report"...

File: pkg/storaged/nfs-mounts.py (hunk 4 above)
Comment (mvollmer) [reply pkg/storaged/nfs-mounts.py]: > Why does it need to happen in Java? The bridge already provides the inotify ↔ channel implementation, so that could be extended to be able to watch for directories too (what's done in Python). Then JS should only use the file.watch() API.

Hmm, I don't follow.  In this PR, we don't need to watch directories.  Are you thinking of `pkg/apps`?  Here, we need to watch files in `/proc`, which don't work with `inotify` but they do work with `select`.

We can extend the fswatch1 payload type to also do `proc` / `sysfs` style change notifications.  (And we should in any case.)

But don't we then need to write all the logic that is now in `nfs-mounts.py` in JavaScript and run it in the browser?  Or do you want to do the file watching in the browser, and the fstab parsing etc in Python?

(I actually started out with implementing this in JavaScript, when I thought I don't need to watch `/proc/self/mounts`, but I quickly got too irritated.)

File: pkg/storaged/nfs-mounts.py (hunk 5 above)
Comment (mvollmer) [reply pkg/storaged/nfs-mounts.py]: Hmm, there doesn't seem to be a way to update an entry in place, or remove one.  In fact, udisks2 also does its own parsing and unparsing of /etc/fstab.

File: pkg/storaged/nfs-mounts.py (hunk 4 above)
Comment (martinpitt) [reply pkg/storaged/nfs-mounts.py]: The inotify interface looked pretty much the same, so that should be split out. And you just did, thanks! AFAICS you still use `select()` on an inotify fd (which is the standard way how to do it), so it doesn't work fundamentally differently than in apps (there it just uses a blocking `read()` on the inotify socket, but there is no reason to not use `select()` there also). Not having to watch directories is a difference, I missed that.

So I still wonder how inotify-watching a file in /proc fundamentally differs from watching a file in /usr? And could that not be done much more cleanly in the bridge (in the future)? Then nfs-mounts.py would just do the actions, not the watching? These two parts should be reasonably independent from each other, aren't they?

The "action" parts don't seem to be much more difficult in JS than in Python at first sight, but you worked on this much more closely. I don't mind a Python shim if that makes things easier.

So in summary, I'm happy with the watching in Python as an initial solution (particularly now that the duplicate code got refactored), I'm just thinking aloud how this should look in the future.

File: tools/cockpit.spec (hunk 11 above)
Comment (mvollmer) [reply tools/cockpit.spec]: Done.

(There is also `cockpit-docker` that depends on `python2`.  I guess that needs cleaning up, right?)

File: pkg/storaged/nfs-mounts.py (hunk 4 above)
Comment (mvollmer) [reply pkg/storaged/nfs-mounts.py]: > So I still wonder how inotify-watching a file in /proc fundamentally differs from watching a file in /usr?

In a nut shell, files in `/proc` and `/sys` don't ever generate inotify events.  See for example http://www.spinics.net/lists/linux-fsdevel/msg73924.html

File: tools/cockpit.spec (hunk 11 above)
Comment (martinpitt) [reply tools/cockpit.spec]: No, cockpit-docker already [has](https://github.com/cockpit-project/cockpit/blob/master/tools/debian/control#L89) `Depends: python3`, and `debian.setup` does not have any python package. So that should be fine.

File: pkg/storaged/nfs-mounts.py (hunk 4 above)
Comment (martinpitt) [reply pkg/storaged/nfs-mounts.py]: Indeed I was missing `select.select([ fstab_watcher.inotify.fd ], [ ], [ mtab_file ])`, which watches those separately. So yes, using the python wrapper for this bit seems fine for now. Thanks for bearing with me!

**Comment by mvollmer** [comment]
This failure looks genuine:

`
Unexpected journal message 'stat: cannot read file system information for '/mounts/bar': No such file or directory'
`

**Comment by mvollmer** [comment]
> stat: cannot read file system information for '/mounts/bar': No such file or directory

This happens almost every time when changing the mount point.  I have fixed this by changing the order in which the changes are made (so that Cockpit doesn't see a state that will soon be invalid), and by keeping the error message out of the journal (since it is harmless).

**Comment by andreasn** [comment]
I noticed a strange behaviour where if I entered something wrong (like the wrong path on the server), it would give me the error (correctly), but mount the NFS share anyway.

**Comment by mvollmer** [comment]
> I noticed a strange behaviour where if I entered something wrong (like the wrong path on the server), it would give me the error (correctly), but mount the NFS share anyway.

Yes, good catch.  I found a whole lot of little things to fix in case of errors.

**Comment by mvollmer** [comment]
@martinpitt, could you have another look?

File: pkg/storaged/index.html (hunk 2 above)
Comment (mvollmer) [reply pkg/storaged/index.html]: Good catch!  5 lines down shoudl also use "data-action".  Now the dialog code needs to do some clean up when it is closed, so we need to put a real handler on the Cancel and Close buttons.

**Comment by martinpitt** [comment]
I re-reviewed, thanks for the fixes! Outstanding issues:
 - https://github.com/cockpit-project/cockpit/pull/7580#discussion_r143161861 (remaining `data-dismiss` need to be renamed)
 - https://github.com/cockpit-project/cockpit/pull/7580#discussion_r143161131 (don't continuously poll df on NFS)

Other than that, looks good to me!

File: pkg/storaged/index.html (hunk 2 above)
Comment (mvollmer) [reply pkg/storaged/index.html]: Fixed.

File: src/bridge/cockpitmountsamples.c (hunk 1 above)
Comment (mvollmer) [reply src/bridge/cockpitmountsamples.c]: Fixed.  Now we poll only once per fstab change.

**Comment by mvollmer** [comment]
> Outstanding issues:

Both should be fixed.

**Comment by mvollmer** [comment]
Ahh, clicked 10 pixels to far to the left!  Sorry.

**Comment by martinpitt** [comment]
Voilà, nice and green!
